### ISSUE TITLE
Use cudf byte ranges to build proper read_csv

### ISSUE THREAD

**Issue by mrocklin** [rapidsai/dask-cudf#71]
Currently the read_csv function reads each file as a single dataframe.  This can be a problem when we have larger files.  Now that cudf supports byte ranges we should be able to read large files into many small blocks (see https://github.com/rapidsai/cudf/pull/807)

This would allow for smoother operation on large files.

For each file we need to look at the size of the file, then create tasks to read each chunk of data from that file, being mindful of when to read and not read the header.  We'll need to accept a `blocksize=` keyword as well (or whatever `dask.dataframe.read_csv` uses).

Eventually it would be nice to unify this code with dask.dataframe's implementation, but I suspect that that is too complex to get done in the near term, so a short term fix might be appropriate.

**Comment by mrocklin** [comment]
cc @quasiben

**Comment by quasiben** [comment]
@mrocklin are you suggesting reimplementing the logic in [text_block_to_pandas](https://github.com/dask/dask/blob/master/dask/dataframe/io/csv.py#L230-L247) and more in `dask.dataframe.read_csv` for `dask-cudf` ? Or would this be flushing out the `(apply, cudf.read_csv, [fn], kwargs)` in the graph ? Perhaps you are thinking of something else entirely

**Comment by mrocklin** [comment]
Something like the former.  We currently build a graph that is one-task-per-file.  We would like a graph that is one-task-per-chunk-of-file.  The dask.dataframe code is particularly mature.  I don't think we need to go that far (the code in this repository will hopefully someday be just replaced by that code).  I would imagine that this would look something like the following:

```python
dsk = {}
i = 0
for fn in sorted(glob(path)):
    size = os.get_file_size_of(fn)
    for start in range(0, size, chunksize):
        kwargs2 = kwargs.copy()
        kwargs2['byte_range'] = (start, chunksize)  # specify which chunk of the file we care about
        if start != 0:
            kwargs2['header'] = False  # no header in the middle of the file
        dsk[(keyname, i)] = (apply, cudf.read_csv, [fn], kwarg2)
        i += 1
```

Though there are many things above that I've probably glossed over.  Looking through the dask.dataframe code is a good way to get a sense of the issues that have cropped up before.  However some subjective judgement will need to be taken to determine which of these issues we should think about short term, and which we should just punt on in order to get something out somewhat quickly.  Again, hopefully this particular piece of code is short-lived (though who knows).

**Comment by mrocklin** [comment]
For people watching this this is implemented in #78 and works fine.  However it has uncovered a couple issues upstream where we drop the occasional row (see https://github.com/rapidsai/cudf/issues/946)

If people like @randerzander or @zronaghi have time to try out that branch that would be welcome.

**Comment by zronaghi** [comment]
I just tested that branch on a csv file that we discussed and works fine, thanks!

**Comment by randerzander** [comment]
Closed by #78
